### PR TITLE
[SPARK-35361][SQL] Improve performance for ApplyFunctionExpression

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

In `ApplyFunctionExpression`, move `zipWithIndex` out of the loop for each input row.

### Why are the changes needed?

When the `ScalarFunction` is trivial, `zipWithIndex` could incur significant costs, as shown below:

Screen Shot 2021-05-11 at 10 03 42 AM

By removing it out of the loop, I'm seeing sometimes 2x speedup from `V2FunctionBenchmark`. For instance:

Before:
```
scalar function (long + long) -> long, result_nullable = false codegen = false:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
native_long_add                                                                         32437          32896         434         15.4          64.9       1.0X
java_long_add_default                                                                   85675          97045         NaN          5.8         171.3       0.4X
```

After:
```
scalar function (long + long) -> long, result_nullable = false codegen = false:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
native_long_add                                                                         30182          30387         279         16.6          60.4       1.0X
java_long_add_default                                                                   42862          43009         209         11.7          85.7       0.7X
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ApplyFunctionExpression.scala [hunk 1]
```diff
@@ -30,10 +30,11 @@ case class ApplyFunctionExpression(
   override def dataType: DataType = function.resultType()
 
   private lazy val reusedRow = new GenericInternalRow(children.size)
+  private lazy val childrenWithIndex = children.zipWithIndex
 
   /** Returns the result of evaluating this expression on a given input Row */
   override def eval(input: InternalRow): Any = {
-    children.zipWithIndex.foreach {
+    childrenWithIndex.foreach {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ApplyFunctionExpression.scala:37]: shall we simply use a while loop with a counter? Adding a lazy val as a class member variable looks a bit heavy.

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ApplyFunctionExpression.scala:37]: I'm not sure if this is very useful. After the change the cost of the loop is negligible:

Screen Shot 2021-05-11 at 10 56 15 AM

Plus I think `foreach` is easier to read.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ApplyFunctionExpression.scala:37]: Actually @sunchao it's in the style guide: https://github.com/databricks/scala-style-guide#traversal-and-zipwithindex. Looks like here's a hot path.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ApplyFunctionExpression.scala:37]: I think it's worthwhile trying.

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ApplyFunctionExpression.scala:37]: Sure I can give it a try later if you have strong opinion on this.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ApplyFunctionExpression.scala:37]: thanks man!

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ApplyFunctionExpression.scala:37]: Running benchmarks now. Let's see.

**Pull request closed without merging**